agent: reduce directory errors, easier debugging, various improvements

dpausp opened the pull request:
A collection of commits to reduce the need for sudo for non-invasive fc-maintenance commands, improve error handling, logging, avoiding unneccessary directory API calls and retrying failed calls. See commit messages for details.

@flyingcircusio/release-managers

## Release process

Impact:

Changelog:

 * `fc-maintenance` commands for viewing requests (`show` and `list`) can now be used without sudo by all users. The same applies to the commands used for monitoring and telemetry (`check` and `metrics`).
 * `fc-maintenance show <request ID prefix>`: now has improved output and also works for archived maintenance requests while active requests are preferred when there are multiple matches.
 * Command output generated during execution of an update activity is now logged to separate files in `/var/log/fc-agent`, like for `fc-manage` calls.
* Unhandled errors in `fc-manage` and `fc-maintenance` are now logged properly to the system journal and can be viewed with `journalctl SYSLOG_IDENTIFIER=fc-agent`.

## Security implications

This change affects sudo rules for fc-maintenance commands but the set of commands that can be run and the allowed groups have not been extended.

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - reduce need to use sudo for commands that only access data readable by any user  
  - make sure that invasive commands cannot be run by normal users without sudo and that sudo permissions are limited properly (admin, sudo-srv and service may only run fc-maintenance (-v) delete without password)
- [x] Security requirements tested? (EVIDENCE)
  - checked on a test VM that fc-maintenance commands work as intended, checked sudo rules, checked logs
  - automated test have been extended

dpausp committed:
agent: only schedule requests in State.due

Scheduling requests that are already running or are in a state
to be archived isn't useful and may cause unneccessary errors.

PL-131813

dpausp committed:
agent: improve shell script activity rendering

PL-131813

dpausp committed:
agent: command logging for update activity runs, ignore main log file permission errors

Output of nix commands in UpdateActivity.run() is now logged. A typical
problem are failing services on system activation which we want to be
logged to a separate file like for `fc-manage switch`.

Main log file permissions problems are now ignored.
No need to abort when the main log file cannot be used.
We still have journal logging and console output.

PL-131813

dpausp committed:
agent: fc-maintenance allows non-invasive commands for non-root

We now separate non-invasive and invasive code paths better.
Moves around existing methods to a more logical order, grouping
invasive and non-invasive methods.

Some read-only commands for showing requests, metrics and the Sensu
check can now be called by non-root users without the need for sudo.

Clean up some uses of rm.scan() which are now handled by
__enter__ which must be called for all invasive methods. This also
takes care of loading requests and creating missing directory now.

Add missing @require_lock decorators for invasive methods and give
internal methods a underscore prefix. The latter don't have the
decorator but it should be fairly obvious how to use them.

PL-131813

dpausp committed:
agent: retry directory API calls

Use the `stamina` library for automatic retries with
integrated logging, exponential backoff and jitter.

PL-131813

ctheune commented on pkgs/fc/agent/fc/util/directory.py:
```diff
@@ -18,6 +20,40 @@ def __init__(self, url, *args, **kw):
+        handler = self._ServerProxy__handler
+        return f"Directory API ({host}{handler})"
+
+
```
retry behaviour should be customizable here. default, based on our code base, IMHO should be `off`

ctheune reviewed (commented):
Looks good in general, see the small comment about making retry behaviour configurable.

dpausp committed:
agent: move reboot check from load to run in VMChangeActivity

This has some advantages:

* Dmidecode as external tool is called less often, only when really
  needed (init, just before reboot). Before, this happened on every
  agent run.
* Less debug log messages.
* Activity doesn't change just by loading it.
* Non-privileged users can show the activity now.

PL-131813

dpausp committed:
agent: capture output from enter commands, handle exceptions

Exceptions from _enter_maintenance don't bubble up anymore but are
logged and treated like temporary failures now.

Output from enter commands is now shown directly on the trace log level
and added to the exception if the command fails or logged when
postpone/tempfail is requested.

PL-131813

dpausp committed:
agent: only add pyslurm dependency on Linux

We cannot/don't want to build pyslurm on MacOS.

PL-131813

dpausp committed:
agent: log unhandled exceptions in commands

Before, exceptions just bubbled up to Typer's excepthook which
pretty-prints them. This is nice for interactive use but not for agent
tasks run by systemd units. They ended up in stdout while other log
messages are send to journal directly. This means that exceptions had a
different SYSLOG_IDENTIFIER than log messages which is annoying for
debugging.

Also, we didn't log unhandled exceptions that occurred in
interactive use at all.

This change adds a new FCTyperApp which is used for fc-manage and
fc-maintenance commands. The class extends Typer's app class and adds
exception logging. It still passes exceptions to Typer's excepthook when
used interactively.

PL-131813

dpausp committed:
agent: fix default value for datetime comparison

We need to use timezone-aware objects here.

PL-131813

osnyx commented on pkgs/fc/agent/fc/maintenance/tests/test_cli.py:
```diff
@@ -43,9 +43,11 @@ def app_main_args(tmp_path, agent_maintenance_config):
+def invoke_app(app_main_args, monkeypatch):
     runner = typer.testing.CliRunner()
 
+    monkeypatch.setattr("os.getuid", lambda: 0)
```
As this now causes all tests with that fixture to be run as root, it might make sense to codify this in the fixture name.
I understand that this simplifies most of the test cases or is even explicitly intended. Yet, we need to be careful to always be aware of this when writing test cases where the invoking user is relevant.
I see that in `test_invoke_metrics` you explicitly do __not__ use this fixture, as running it as non-root is the intended case there.

osnyx commented on pkgs/fc/agent/default.nix:
```diff
@@ -34,6 +35,20 @@ let
+    };
+
+    nativeBuildInputs = with py; [ hatchling hatch-vcs hatch-fancy-pypi-readme ];
+    propagatedBuildInputs = with py; [ structlog tenacity typing-extensions ];
```
Just a note, as this confused me: We rely on name scope shadowing here, as pkgs.tenacity [is a graphical audio editor](https://tenacityaudio.org/) instead of the python retry library.
Totally fine to do so, it can be a matter of taste to make this distinction clear by still explicitly writing `py.tenacity`.

dpausp replied:
I think it's ok here because `with py;` is close to the place where the package is used and adding another explicit `py.tenacity` looks strange . In general, qualifying packages is a good idea when packages from multiple sources are mixed in a list.

osnyx commented on pkgs/fc/agent/fc/maintenance/reqmanager.py:
```diff
@@ -640,7 +766,7 @@ def _handle_enter_tempfail(
                 _replace_msg=(
                     "Run all mode requested but a maintenance enter "
-                    "command had a temporary failure."
+                    "command had a temporary) failure. "
```
Typo? I am missing the opening bracket.

osnyx reviewed (changes requested):
LGTM overall, but with a few minor changes.
Also this is still a draft, requiring fixing the pre-commit failures and the usual PR documentation.

dpausp committed:
agent: fix parsing of oneline error msgs

Also logs the number of error lines to parse and the last 25 lines
if parsing fails.

PL-131813
